ally/scraper.py

```diff
@@ -0,0 +1,171 @@
+"""
+Get Ally Transactions using Selenium
+"""
+from datetime import datetime
+import json
+import os
+import sys
+
+import psycopg2.extras
+from selenium import webdriver
+from selenium.webdriver.common.by import By
+from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+
+
+ALLY_URL = 'https://www.ally.com/'
+
+
+def find_element_by_attribute(driver, tag_name, attribute, value):
+    elements = driver.find_elements_by_tag_name(tag_name)
+    for element in elements:
+        val = element.get_attribute(attribute)
+        if val == value:
+            return element
+
+
+class AllyScraper:
+
+    def __init__(self, driver, username, password, account):
+        self.driver = driver
+        self.username = username
+        self.password = password
+        self.account = account
+
+    def login(self):
+        # Click login
+        self.driver.get(ALLY_URL)
+        self.driver.find_element_by_id('login-btn').click()
+
+        WebDriverWait(self.driver, 10).until(
+            EC.visibility_of_element_located((By.ID, 'login-widget'))
+        )
+
+        # Select account type
+        self.driver.find_element_by_xpath("//select[@id='account']/option[text()='Bank or Invest Login']").click()
+
+        username_el = self.driver.find_element_by_id('username')
+        username_el.send_keys(self.username)
+
+        password_el = self.driver.find_element_by_id('password')
+        password_el.send_keys(self.password)
+
+        submit_el = find_element_by_attribute(self.driver, 'button', 'data-id', 'submit')
+        submit_el.click()
+
+    def goto_account_page(self):
+        acct_containers = self.driver.find_elements_by_class_name('accounts-individual-container')
+
+        for container in acct_containers:
+            spans = container.find_elements_by_tag_name('span')
+            for span in spans:
+                if '...%s' % self.account in span.text:
+                    link = WebDriverWait(container, 10).until(
+                        EC.element_to_be_clickable((By.TAG_NAME, 'a'))
+                    )
+                    link.click()
+                    return
+
+        raise Exception('Account %s not found', self.account)
+
+    def get_transactions(self):
```

Comment:
Consider only returning transactions that aren't already in database

Reply:
done

Reply:
done